#!/usr/bin/env python

import logging
import multiprocessing
import pathlib


from orfaqs.apps.common.cliutils import CliUtil
from orfaqs.apps.common.orfaqsapp import ORFaqsApp
from orfaqs.apps.common.orfaqsproteindiscovery import (
    ORFaqsProteinDiscoveryUtils,
)

from orfaqs.lib.utils.directoryutils import DirectoryUtils


_logger = logging.getLogger(__name__)


class ORFaqsProteinDiscovery(ORFaqsApp):
    """ORFaqsProteinDiscovery"""

    @staticmethod
    def program_name():
        return 'ORFaqs Protein Discovery'

    @staticmethod
    def cli():
        try:
            arg_descriptor_list = [
                CliUtil.create_new_arg_descriptor(
                    ('-i', '--input_sequence'),
                    arg_help=(
                        'A sequence file or genomic sequence string. '
                        'Currently, only FASTA files are supported.'
                    ),
                    arg_type=str,
                ),
                CliUtil.create_new_arg_descriptor(
                    ('-j', '--job_id'),
                    arg_help=(
                        'A string used to uniquely identify a set of '
                        'results generated by the tool. All results '
                        'with the same job_id share the same directory '
                        'sub-folder.'
                    ),
                    arg_type=str,
                    default=None,
                ),
                CliUtil.create_new_arg_descriptor(
                    ('-o', '--output_directory'),
                    arg_help=(
                        'The output directory for exported files '
                        'and results. If the path does not exist, one '
                        'will be created. The default directory is: '
                        f'{ORFaqsProteinDiscovery.default_output_directory()}'
                    ),
                    arg_type=str,
                    default=ORFaqsProteinDiscovery.default_output_directory(),
                ),
                CliUtil.create_new_arg_descriptor(
                    ('--export_format'),
                    arg_help=(
                        'Defines the exported format of the resulting data. '
                        'Supported formats are: '
                        f'{ORFaqsProteinDiscoveryUtils.available_export_formats()}'
                    ),
                    arg_type=str,
                ),
                CliUtil.create_new_arg_descriptor(
                    (f'--{ORFaqsProteinDiscovery.launch_json_option_name()}'),
                    arg_help=(
                        'A JSON file specifying arguments and options to '
                        "apply to the program's session. If "
                        'arguments provided in the JSON overlap with '
                        'arguments provided on the command line, values from '
                        'the command line override those specified in the JSON.'
                    ),
                    arg_type=str,
                ),
            ]

            cli_arg_parser = CliUtil.create_arg_parser(
                arg_descriptor_list,
                program_name=ORFaqsProteinDiscovery.program_name(),
                description=(
                    'Discover protein sequences from DNA or RNA sequences.'
                ),
                epilog='',
            )

            # Process the CLI args and incorporate any
            # values set in the launch json.
            ui_kwargs = ORFaqsProteinDiscovery.process_ui_kwargs(
                **CliUtil.parse_args(cli_arg_parser)
            )
            if 'input_sequence' not in ui_kwargs:
                message = (
                    '[ERROR] No input file or sequence string was '
                    'specified. A valid input is required.'
                )
                _logger.error(message)
                print(message)
                cli_arg_parser.print_help()
            else:
                ORFaqsProteinDiscovery.run(**ui_kwargs)

        except SystemExit as e:
            if e.args[0] is True:
                raise SystemExit from e

    @staticmethod
    def _run(
        input_sequence: str | pathlib.Path,
        output_directory: str | pathlib.Path = None,
        job_id: str = None,
        export_format: str = None,
        **kwargs,
    ):
        if output_directory is None:
            output_directory = (
                ORFaqsProteinDiscovery.default_output_directory()
            )

        # Create the local output directory and output file path
        output_directory = DirectoryUtils.make_path_object(output_directory)
        if isinstance(job_id, str):
            output_directory = output_directory.joinpath(job_id)

        if DirectoryUtils.is_file(input_sequence):
            # Try processing as a FASTA file
            ORFaqsProteinDiscoveryUtils.process_fasta_file(
                fasta_file_path=input_sequence,
                export_format=export_format,
                output_directory=output_directory,
            )
        else:
            # Try processing as an sequence string.
            ORFaqsProteinDiscoveryUtils.process_genomic_sequence(
                genomic_sequence=input_sequence,
                export_format=export_format,
                output_directory=output_directory,
            )


if __name__ == '__main__':
    multiprocessing.freeze_support()
    ORFaqsProteinDiscovery.cli()
